### ISSUE TITLE
Reorganization of attributes in the event model classes

### ISSUE THREAD

**Issue by paolafer** [next-exp/IC#586]
@gonzaponte and I were talking about the organization of the attributes in the classes of the event model, such as `BHit` and derived, but not only those.
I find difficult to understand why some attributes have `@properties` (which don't allow to modify the attribute) but we actually modify them in the usage, or why some are marked with an underscore and some aren't etc. etc. 
I propose to adopt some consistent convention, so that it will be easier to implement new classes or modify the existing ones to future developers.
My proposal would be to add a `@property` only to those attributes that we consider unchangeable and don't want to be modified by the user. For instance, the energy of a `BHit` object is not the same forever, since we correct it, so I would eliminate the `E` property. As for the position, there is at least a case where it's changed, because it's smeared (Javi's fast analysis algorithms), so I would also eliminate the property. In the case of position we could maybe have `x`, `y` and `z` set as attributes with no underscores and a `pos` property that returns a value that is calculated from `x`, `y` and `z`.
Opinions?

**Comment by jjgomezcadenas** [comment]
I actually have a more radical proposal. I would like to use only dataclasses in the future. This is the approach that I have used in the KrCalib software. 

Dataclasses are intended to provide "types" which can be manipulated by functions. An example: 
```
@dataclass
class CPoint:
    """Represent a cartesian point"""
    X   : np.array
    Y   : np.array
    Z   : np.array
```

Notice that you only declare the attributes of the object (and declare its type, which is useful). Constructors, streaming and such are automatically provided by the dataclass. Access is by name (as in the case of namedtuple). Unlike namedtuples, dataclasses provide inheritance, thus, for example:

```
@dataclass
class Point(CPoint):
    """Add polar representation"""
    R   : np.array
    Phi : np.array


@dataclass
class KrEvent(Point):
    """Adds raw energy/time"""
    S2e  : Array
    S1e  : Array
    S2q  : Array
    T    : Array  # time
    DT   : Array  # time difference in seconds
    E    : Array
    Q    : Array

```

(if you are curious Array is just a defined name for np.array). Dataclasses are mutable, although I believe we should try to use them as inmutable objects. Their use simplify enormously the definition of types and encourages manipulating such types with functions. 

For example, our current class:
```
class SensorParams:
    """Transient class storing sensor parameters."""
    def __init__(self, npmt, pmtwl, nsipm, sipmwl):
       self.npmt   = npmt
       self.pmtwl  = pmtwl
       self.nsipm  = nsipm
       self.sipmwl = sipmwl

    @property
    def NPMT  (self): return self.npmt

    @property
    def PMTWL (self): return self.pmtwl

    @property
    def NSIPM (self): return self.nsipm

    @property
    def SIPMWL(self): return self.sipmwl

    def __str__(self):
       s = "{0}SensorParams\n{0}".format("#"*20 + "\n")
       for attr in self.__dict__:
           s += "{}: {}\n".format(attr, getattr(self, attr))
       return s

    __repr__ = __str__
```
Could be rewritten as dataclass as:

```
@dataclass
class SensorParams:
 NPMT     : int
PMTWL   : np.array
NSIPM     : int
SIPMWL  : np.array

```
Much more concise and to the point, and avoiding implementation details (such as the use or not of @property). 

Beyond the use of dataclasses I would like to move towards a more functional style approach. In particular trying to avoid side effects. A classical side effect is to send an instance of a class to a function that modifies it in situ so that after the function the object has changed in an unpredictable way.  Example:

```
a = ['hello']

def side_effects(a):
    a.append('baby')
    return 'I love side effects'

print(a)
b = side_effects(a)
print(a)
print(b)

```

Run it to get:

```
['hello']
['hello', 'baby']
I love side effects
```

a function without side effects would do:

```
from typing      import List
a = ['hello']

def typehinted_with_no_side_effects(a : List) ->List:
    b = list(a)
    b.append('baby')
    return b

print(a)
b = typehinted_with_no_side_effects(a)
print(a)
print(b)
```

and now:
```
['hello']
['hello']
['hello', 'baby']
```

Notice that in the previous example nothing guarantees that the function does not have side effects because we are passing a mutable type (a list). To be sure that the function will not have side effects we could instead pass a tuple, which is immutable. Dataclasses can also be made to behave as immutable objects. See discussion (and nice tutorial) here:

https://realpython.com/python-data-classes/


To summarize, I would propose a serious rethinking of our data model based on dataclasses and keeping in mind the above discussion. I volunteer to take the lead here.

**Comment by paolafer** [comment]
Ok, I like it!

**Comment by jahernando** [comment]
I also like the dataclass concept. 

In the example above, the str method is in fact a general function which can apply to all data classes.

There is one question more. It looks to me that Voxel could be a Hit composite, if we need to access a collection of hits inside, or event just a Hit. Hit and Voxel could be the same thing. Maybe we are complicated ourselves with the two classes.

**Comment by paolafer** [comment]
I disagree on this last point. A Voxel  can still inherit from a Hit, as currently does, but we need to keep them separate to account for the hits that are used to build a voxel. This is a piece of information that we're using right now.

**Comment by jacg** [comment]
Re: @jjgomezcadenas dataclass thesis (let's see how much I can say before the students come back from lunch):

First, a couple of minor (somewhat orthogonal to the main issue) points.

1. You use docstrings as if they were comments. Please don't. Take the docstring in you class `Point`. A user who wants to know what `Point` is about might type `Point?` into an IPython propmt and get told "Add polar representation". This just makes no sense whatsoever. **Docstrings are NOT comments**.

2. What's with the capital letters in all over the place? It violates both PEP8 and standard mathematical/physical notational conventions. `R`, `Phi` `X`, `Y`, `Z`, `T` should all be all-lowercase.

On to the important points. I get the impression that your infatuation with dataclasses and functional programming is blinkering your view and making you fail to see some improtant issues. Don't get me wrong: You know that I think that functional programming is great, that I want to see more of it in general and in IC in particular, and that I really am happy that dataclasses were added to the language.

Consider

    @dataclass
    class CPoint:
        """Represent a cartesian point"""
        X   : np.array
        Y   : np.array
        Z   : np.array

    @dataclass
    class Point(CPoint):
        """Add polar representation"""
        R   : np.array
        Phi : np.array

There are 5 variables here but only 3 degrees of freedom. You need to introduce a constraint, so that construction of Point ensures that `x`, `y`, `r` and `phi` (I have taken the liberty to use orthodox names), which means writing your own constructor, which means that one of the advantages of dataclasses (automatically writing constructors for you) disappears. Now you have the problem of maintiaining this consistency as a class invariant. This wouldn't be an issue if instances were immutable, but, as you noted, they are not. The obvious way of maintaning this consistency is with properties. Bang goes another advantage of dataclasses: you can't use the dataclass magic to define `r` and `phi`.

That's all I have time for, for now. Let me summarize by saying that, from my perspective of seeing these things go past without having the time to look at them carefully, I'm starting to get the impression that you've discovered the hammer and now you think that everything is a nail ... and I'm starting to worry just a little bit that when I next have time to look at IC I will discover that it's been replaced with a sack of nails.

**Comment by jjgomezcadenas** [comment]
> You use docstrings as if they were comments. Please don't. Take the docstring in you class Point. A user who wants to know what Point is about might type Point? into an IPython propmt and get told "Add polar representation". This just makes no sense whatsoever. Docstrings are NOT comments.

Got it. 

>What's with the capital letters in all over the place? It violates both PEP8 and standard mathematical/physical notational conventions. R, Phi X, Y, Z, T should all be all-lowercase.

Yeah, well, we do this somethimes, there is a bit of all of that spread over IC. For the record, the examples I was discussing are from KrCalib (not IC). Yes, it should all be lower case. 

>There are 5 variables here but only 3 degrees of freedom. You need to introduce a constraint, so that construction of Point ensures that x, y, r and phi (I have taken the liberty to use orthodox names), which means writing your own constructor, which means that one of the advantages of dataclasses (automatically writing constructors for you) disappears. Now you have the problem of maintiaining this consistency as a class invariant. This wouldn't be an issue if instances were immutable, but, as you noted, they are not. The obvious way of maintaning this consistency is with properties. Bang goes another advantage of dataclasses: you can't use the dataclass magic to define r and phi.

Actually, I don't need to write my own constructor, something of the type p = Point(x,y,z,r,phi) will work. Perhaps I have missed your point. 

>On to the important points. I get the impression that your infatuation with dataclasses and functional programming is blinkering your view and making you fail to see some improtant issues. 

Well here is the master complaining to the student. Got a reprimand when I was a fun of objects, got another one now for being in love with functions. Man, I (perhaps we all) often fail to see important issues (even you some time fail to see important, rather than improtant). Yet, I think that the general ideas are quite sound. 

>The impression that you've discovered the hammer and now you think that everything is a nail 

Perhaps so. Did you think that every city was a River? We try as we learn.

>next have time to look at IC I will discover that it's been replaced with a sack of nails.

Well, there is not a single dataclass in IC yet, all my experimentation is in KrCalib, among other things to learn the ropes. As for IC, each nail will be in a PR, which will be reviewed before accpeting it. I guess you agree we cannot simply sit down and wait until you have time to check... 

So far I have not understood the criticism beyond a couple of trivial points (docstrings, capital letters), and an observation regarding the constructors that I haven't quite grasped. I have not seen an example that justifies why my infatuation (is that really the correct way to put it?) with functional programming (which I learned from you) or datalcasses is leading IC to nail disaster...

**Comment by jjgomezcadenas** [comment]
Oh, after my long wailing a I see your point. Of course the example I chose was a stupid one, given the fact that R and Phi can be computed from X,Y,Z. It has nothing to do with dataclasses, though (you can add methods to datalclasses if you want). The reason I defined such object was that the kdst produced by DOROTHEA includes X,Y,Z,R,PHI (btw with non kosher names) and so I would read the five values directly from the kdst (e.g, no need to compute R and PHI). 

So, not much to do with infatuation, and more related with the usual incompetence. And yet, I don't think we are nailing IC's coffin here...

**Comment by jacg** [comment]
I think you're starting to see the seed of my important point, which is:

+ @paolafer asks about properties
+ @jjgomezcadenas replies with something along the lines of "Never mind properties, look at how amazing dataclasses are" and proceeds to back it up with the poster-child example of a situation where properties should be used, and dataclasses are actually a pretty poor fit.

Specifically: yes, dataclasses write constructors for you, but in the example given, the constructor written by the dataclass is worse than useless, as it is fatally flawed. To be clear

> Actually, I don't need to write my own constructor, something of the type p = Point(x,y,z,r,phi) will work. 

`Point(x,y,z,r,phi)` shoudn't be allowed, because the user could then say `Point(x=0, y=0, z=0, r=666, phi=0)` which is self-contradictory (because `x = y = z = 0` implies that `r = 0`). This is a classical situation in which properties are the correct approach; the dataclass-generated constructor is wrong.

Incidentally, you could get dataclasses and properties to cooperate effectively in this sort of situation:

    from math        import sqrt
    from dataclasses import dataclass
    
    @dataclass
    class Point:
    
        x : float
        y : float
    
        @property
        def r(self):
            x, y = self.x, self.y
            return sqrt(x*x + y*y)
    
    p = Point(3,4)
    
    assert p.r == 5

> Got a reprimand when I was a [fan] of objects, got another one now for being in love with functions.

You're not being reprimanded: I'm just gently pointing out that not everything is a nail, and that there are other tools in the toolbox which might be more appropriate in some situations. Specifically @paolafer asked about our policy for can-opener use, and you replied by showing how your shiny new hammer can be used to open cans. [Hint: it causes a mess!]

I *did* explicitly say in my original comment that I want to see more functional programming (and dataclasses) in IC. I encourage you to pursue these ideas, but I also encourage you to discriminate between situations in which whey work well and ones in which work less well. Your example suggested that your discrimination-routines need fixing and your enthusiams could do with a little tempering.

> As for IC, each nail will be in a PR, which will be reviewed before accpeting it.

... and given the amount of time I have to spare, I will probably not be involved in that reviewing process, so I need to get my warnings in before then.

> I guess you agree we cannot simply sit down and wait until you have time to check...

Of course I agree. There might not even be any of me left to wait for, so you absolutely shouldn't wait for me. I would like to be involved in the discussion and exploration, but I'm simply not available for that any time soon.

**Comment by jjgomezcadenas** [comment]
>Point(x,y,z,r,phi) shoudn't be allowed, because the user could then say Point(x=0, y=0, z=0, r=666, phi=0) which is self-contradictory (because x = y = z = 0 implies that r = 0). This is a classical situation in which properties are the correct approach; the dataclass-generated constructor is wrong.

Actually is a bit trickier than that, because the Point object is used to capture information that comes from the kdst. And the kdst includes X,Y,Z,R,PHI. So, in a sense this is God given info we don't want to recompute (and we need). So actually we should allow the constructor and explain why, I guess.


>You're not being reprimanded:

I know, I know. It was the drama Queen in me, asking for some limelight. 

>so I need to get my warnings in before then.

Got it. Actually, I think I was careless showing the example, but there is nothing wrong in the dataclass, perhaps the field should be called frozenR and frozenPhi, meaning that it takes something passed from outside. 

>There might not even be any of me left to wait for
Well, let's cross fingers (not sure what to cross fingers for, but hope for the best in any case).

**Comment by mmkekic** [comment]
Done in #643
